presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
+                    groupStageStatesPerTask.get(tableCommitContext.getTaskId()).update(page, true);
+
+                    if (isLastPageForGroupedLifespan(page)) {
+                        checkState(!finishedRecoverableLifespanStages.containsKey(lifespanStage), "GroupStage already finished");
+                        LifespanStageState lifespanStageState = groupStageStatesPerTask.get(tableCommitContext.getTaskId());
+                        finishedRecoverableLifespanStages.put(lifespanStage, lifespanStageState);
+                        unfinishedRecoverableLifespanStageStates.remove(lifespanStage);
+                        partitionCommitter.commitPartition(lifespanStage.getLifespan().getId(), lifespanStageState.getFragments());
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unrecognized commit granularity: " + tableCommitContext.getCommitGranularity());
+            }
+        }
+
+        public List<Page> getStatisticsPagesToProcess(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanStage lifespanStage = LifespanStage.fromTableCommitContext(tableCommitContext);
+            switch (tableCommitContext.getCommitGranularity()) {
+                case TABLE:
+                    return extractStatisticsRows(page).map(ImmutableList::of).orElse(ImmutableList.of());
+                case PARTITION:
+                    if (!finishedRecoverableLifespanStages.containsKey(lifespanStage)) {
+                        return ImmutableList.of();
+                    }
+                    return finishedRecoverableLifespanStages.get(lifespanStage).getStatisticsPages();
+                default:
+                    throw new IllegalStateException("Unrecognized commit granularity: " + tableCommitContext.getCommitGranularity());
+            }
+        }
+
+        public long getRowCount()
+        {
+            return Stream.concat(unrecoverableLifespanStageStates.values().stream(), finishedRecoverableLifespanStages.values().stream())
+                    .mapToLong(LifespanStageState::getRowCount)
+                    .sum();
+        }
+
+        public List<Slice> getFragments()
+        {
+            return Stream.concat(unrecoverableLifespanStageStates.values().stream(), finishedRecoverableLifespanStages.values().stream())
+                    .map(LifespanStageState::getFragments)
+                    .flatMap(List::stream)
+                    .collect(toImmutableList());
+        }
+
+        private TableCommitContext getTableCommitContext(Page page)
+        {
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                Block operatorExecutionContextBlock = page.getBlock(CONTEXT_CHANNEL);
+                if (!operatorExecutionContextBlock.isNull(position)) {
```

Comment:
nit: Doesn't context exists in all positions: https://github.com/prestodb/presto/blob/3e0fa7e60409e33af8f0006868157493acffa0ce/presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java#L325-L332

Reply:
You are right. Code here is still in the old world